08-putting-it-all-together.md: rewrite to use random data and draw histograms. (#320)
datacarpentry/python-ecology-lesson: _episodes/08-putting-it-all-together.md

### `plt` pyplot versus object-based matplotlib

Matplotlib integrates nicely with the NumPy package and can use NumPy arrays

> as input of the available plot functions. Consider the following example data,

Reviewer: ... as input *to* ... (?)

Reviewer: sample data

> as input of the available plot functions. Consider the following example data,

Author: ```suggestion
as input to the available plot functions. Consider the following example data,
```

> created with NumPy by drawing 1000 samples from a normal distribution with mean 0 and

Reviewer: a mean value of 0 and a standard deviation of 0.1

> created with NumPy by drawing 1000 samples from a normal distribution with mean 0 and

Author: ```suggestion
created with NumPy by drawing 1000 samples from a normal distribution with a mean value of 0 and
```

> standard deviation 0.1:

Author: ```suggestion
a standard deviation of 0.1:
```

~~~
import numpy

> s = np.random.normal(0, 0.1, 1000)

Reviewer: `np` ~> `numpy` ... or `import numpy as np` but I'd prefer the former.

> s = numpy.random.normal(0, 0.1, 1000)

Reviewer: let's rename `s` to `sample_data`

~~~
{: .language-python}

To make a histogram plot of our draws from the normal distribution, we can use the `hist` function.

> command directly:

Reviewer: > To make a plot of our histogram, we can use matplotib's `hist` function directly:

> To make a histogram plot of our draws from the normal distribution, we can use the `hist` function directly:

Reviewer: `To make a histogram plot` is fine ... but `To plot a histogram` seems simpler, no? Feel free to ignore this suggestion...

Do you think we should emphasize once again that `plt` is `matplotlib.pyplot`?

> we can use the `hist` function provided by `matplotlib.pyplot` directly:

Reviewer: without that clarification, I'm not sure we need `directly` at the end.

Author: I like it because it calls attention to the fact that it isn't `kind=hist`, or similar. But I also don't feel _that_ strongly.

~~~
plt.hist(sample_data)
~~~
{: .language-python}

> ![Line plot of y versus x](../fig/08-normal-distribution.png)

Author: ```suggestion
![Plot of samples from normal distribution](../fig/08-normal-distribution.png)
```

Reviewer: How about
> A histogram of a random sample of 30 draws from a normal distribution with a mean of 0 and a standard deviation of 0.1.

Reviewer: ~30~ 1000

> ![Plot of samples from normal distribution](../fig/08-normal-distribution.png)

Author: ```suggestion
![Histogram of 1000 samples from normal distribution](../fig/08-normal-distribution.png)
```

### `plt` pyplot versus object-based matplotlib

> Matplotlib integrates nicely with the numpy package and can use numpy arrays

Reviewer: I would suggest changing `Numpy` to `NumPy` rather than its lowercase version `numpy`.

as input of the available plot functions. Consider the following example data,

> created with NumPy by drawing samples from a normal distribution:

Author: ```suggestion
created with NumPy by drawing 1000 samples from a normal distribution with a mean of 0 and a sigma of 0.01:
```

Reviewer: I think it's better to spell out `standard deviation` instead of using just `sigma`

~~~
import numpy

> mu, sigma = 0, 0.1

Author: ```suggestion
```

> s = np.random.normal(mu, sigma, 1000)

Reviewer: no need for variables `mu` and `sigma` because they are not used anywhere

Author: ```suggestion
s = np.random.normal(0, 0.1, 1000)
```

Author: Like I've suggested here?

Reviewer: yes. I would suggest adding some extra info to the preceding sentence, something like:
>by drawing <this many> samples from a normal distribution with <such and such> mean value, <such and such> standard deviation

Author: Added this text in a new suggestion above.

~~~
{: .language-python}

> To make a histogram plot of our draws from the normal distribution, we can use the `plot`

Reviewer: Um, you're using `hist` command not `plot` command below

> To make a histogram plot of our draws from the normal distribution, we can use the `plot`

Author: ```suggestion
To make a histogram plot of our draws from the normal distribution, we can use the `hist` function.
```

Reviewer: yes

> To make a histogram plot of our draws from the normal distribution, we can use the `hist` function.

Author: ```suggestion
To make a histogram plot of our draws from the normal distribution, we can use the `hist` function directly:
```

> command directly:

Author: ```suggestion
```

~~~
plt.hist(s)
~~~
{: .language-python}

> ![Line plot of y versus x](../fig/08-normal-distribution.png)

Reviewer: This is no longer a line plot, right?

Reviewer: Another point: if `08_line_plot.png` is no longer used, it shall be removed in the PR or marked as _unused_ by putting it into _`unused`_ folder or something like that...

Author: I'm happy with removal, if it sounds OK to you.

Reviewer: `Line plot of y versus x` is not correct

~~~
fig, ax = plt.subplots()  # initiate an empty figure and axis matplotlib object

> ax.hist(s, 30)

Reviewer: Need to explain `30`

> ax.hist(s, 30)

Author: ```suggestion
ax.hist(s, 30) # 30 is the number of bins, or discrete categories, into which X-axis is broken
```

Reviewer: I think putting it in a standard text would be better. Something like
> where `30` is the number of bins in the histogram.

~~~
{: .language-python}

Although the latter approach requires a little bit more code to create the same plot,
the advantage is that it gives us **full control** over the plot and we can add new items
such as labels, grid lines, title, etc.. For example, we can add additional axes to
the figure and customize their labels:

~~~
fig, ax1 = plt.subplots() # prepare a matplotlib figure
ax1.hist(s, 30)

# Add a plot of a Beta distribution
a = 5
b = 10
beta_draws = np.random.beta(a, b)
# adapt the labels
ax1.set_ylabel('density')
ax1.set_xlabel('value')

# add additional axes to the figure
ax2 = fig.add_axes([0.125, 0.575, 0.3, 0.3])
#ax2 = fig.add_axes([left, bottom, right, top])
ax2.hist(beta_draws)
~~~
{: .language-python}

![Plot with additional axes](../fig/08-dualdistribution.png)

> ## Challenge - Drawing from distributions

> > Have a look at the Numpy

Reviewer: NumPy (for consistency).

> random documentation <https://docs.scipy.org/doc/numpy-1.14.0/reference/routines.random.html>.
> Choose a distribution you have no familiarity with, and try to sample from and visualize it.
{: .challenge}

> 

Reviewer: No need for so many new lines, 2 should be enough

> ## Challenge - Pandas and matplotlib
> Load the streamgage data set with Pandas, subset the week of the 2013 Front Range flood
> (September 11 through 15) and create a hydrograph (line plot) of the discharge data using
> Pandas, linking it to an empty maptlotlib `ax` object. Create a second axis that is the

> > whole dataset. Adapt the title and axes' labels using matplotlib.

Reviewer: I don't understand
> Create a second axis that is the
> whole dataset.

> ## Challenge - Pandas and matplotlib
> Load the streamgage data set with Pandas, subset the week of the 2013 Front Range flood
> (September 11 through 15) and create a hydrograph (line plot) of the discharge data using
> Pandas, linking it to an empty maptlotlib `ax` object. Create a second axis that is the

> > whole dataset. Adapt the title, x-axis and y-axis label using matplotlib.

Reviewer: 1. `labels`.
2. I'd suggest something like:
> Adapt the title and axes' labels using matplotlib

> > whole dataset. Adapt the title, x-axis and y-axis label using matplotlib.

Author: ```suggestion
> whole dataset. Adapt the title, axes labels using matplotlib.
```

Reviewer: > and axes labels using matplotlib.

> ## Challenge - Pandas and matplotlib
> Load the streamgage data set with Pandas, subset the week of the 2013 Front Range flood
> (September 11 through 15) and create a hydrograph (line plot) of the discharge data using
> Pandas, linking it to an empty maptlotlib `ax` object. Create a second axis that is the
> whole dataset. Adapt the title, x-axis and y-axis label using matplotlib.
>
> > ## Answers
> >
> > ~~~
> > discharge = pd.read_csv("../data/bouldercreek_09_2013.txt",
> >                       skiprows=27, delimiter="\t",
> >                       names=["agency", "site_id", "datetime",
> >                              "timezone", "flow_rate", "height"])
> > fig, ax = plt.subplots()

> > > flood = discharge[(discharge["datetime"] >= "2013-09-11") &

Author: Oh, I really like this a lot.

> > > flood = discharge[(discharge["datetime"] >= "2013-09-11") &

Reviewer 2: @wrightaprilm will you update the exercise code or should I do a separate PR?

> > > flood = discharge[(discharge["datetime"] >= "2013-09-11") &

Author: Addressed in 84d5dde - thanks for the reminder!
